// @decorators/express/lib/src/decorators/params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cookies = exports.Headers = exports.Body = exports.Query = exports.Params = exports.Next = exports.Res = exports.Response = exports.Req = exports.Request = void 0;
const meta_1 = require("../meta");
/**
 * Parameter decorator factory, creates parameter decorator
 */
function decoratorFactory(type) {
    return function (name) {
        return function (target, methodName, index) {
            const meta = (0, meta_1.getMeta)(target);
            if (meta.params[methodName] === undefined) {
                meta.params[methodName] = [];
            }
            meta.params[methodName].push({ index, type, name });
        };
    };
}
/**
 * Express req object
 */
exports.Request = decoratorFactory(meta_1.ParameterType.REQUEST);
/**
 * Express req object in short form
 */
exports.Req = exports.Request;
/**
 * Express res object
 */
exports.Response = decoratorFactory(meta_1.ParameterType.RESPONSE);
/**
 * Express res object in short form
 */
exports.Res = exports.Response;
/**
 * Express next function
 */
exports.Next = decoratorFactory(meta_1.ParameterType.NEXT);
/**
 * Express req.params object or single param, if param name was specified
 */
exports.Params = decoratorFactory(meta_1.ParameterType.PARAMS);
/**
 * Express req.query object or single query param, if query param name was specified
 */
exports.Query = decoratorFactory(meta_1.ParameterType.QUERY);
/**
 * Express req.body object or single body param, if body param name was specified
 */
exports.Body = decoratorFactory(meta_1.ParameterType.BODY);
/**
 * Express req.headers object or single headers param, if headers param name was specified
 */
exports.Headers = decoratorFactory(meta_1.ParameterType.HEADERS);
/**
 * Express req.body object or single cookies param, if cookies param name was specified
 */
exports.Cookies = decoratorFactory(meta_1.ParameterType.COOKIES);
