// @decorators/express/lib/src/decorators/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Status = exports.Head = exports.Options = exports.Patch = exports.Delete = exports.Put = exports.Post = exports.Get = exports.All = void 0;
const meta_1 = require("../meta");
/**
 * Route decorator factory, creates decorator
 */
function decoratorFactory(method, url, middleware = []) {
    return (target, key, descriptor) => {
        const methodMetadata = getRouteMeta(target, key);
        methodMetadata.routes.push({ method, url, middleware });
        return descriptor;
    };
}
/**
 * All routes
 *
 * Special-cased "all" method, applying the given route `path`,
 * middleware, and callback to _every_ HTTP method.
 */
function All(url, middleware) {
    return decoratorFactory('all', url, middleware);
}
exports.All = All;
/**
 * Get route
 */
function Get(url, middleware) {
    return decoratorFactory('get', url, middleware);
}
exports.Get = Get;
/**
 * Post route
 */
function Post(url, middleware) {
    return decoratorFactory('post', url, middleware);
}
exports.Post = Post;
/**
 * Put route
 */
function Put(url, middleware) {
    return decoratorFactory('put', url, middleware);
}
exports.Put = Put;
/**
 * Delete route
 */
function Delete(url, middleware) {
    return decoratorFactory('delete', url, middleware);
}
exports.Delete = Delete;
/**
 * Patch route
 */
function Patch(url, middleware) {
    return decoratorFactory('patch', url, middleware);
}
exports.Patch = Patch;
/**
 * Options route
 */
function Options(url, middleware) {
    return decoratorFactory('options', url, middleware);
}
exports.Options = Options;
/**
 * Head route
 *
 */
function Head(url, middleware) {
    return decoratorFactory('head', url, middleware);
}
exports.Head = Head;
/**
 * Method status
 */
function Status(status) {
    return (target, key, descriptor) => {
        const methodMetadata = getRouteMeta(target, key);
        methodMetadata.status = status;
        return descriptor;
    };
}
exports.Status = Status;
function getRouteMeta(target, key) {
    const meta = (0, meta_1.getMeta)(target);
    return meta.routes[key] = meta.routes[key] || {
        routes: [],
    };
}
