// @decorators/di/lib/src/store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Store = void 0;
const reflector_1 = require("./reflector");
const injection_token_1 = require("./injection-token");
class Store {
    /**
     * Get provider id
     */
    static providerId(injectable) {
        if (!injectable) {
            return null;
        }
        if (typeof injectable === 'string' || injectable instanceof injection_token_1.InjectionToken) {
            return injectable;
        }
        return reflector_1.Reflector.getId(injectable);
    }
    /**
     * Register provider
     */
    static provider(type, args) {
        let provider = this.findProvider(type);
        if (provider === undefined) {
            provider = this.createProvider(type);
        }
        if (args === undefined) {
            reflector_1.Reflector.paramTypes(type)
                .forEach((param, index) => {
                if (!provider.deps[index] || !provider.deps[index].id) {
                    provider.deps[index] = { id: this.providerId(param) };
                }
            });
            return type;
        }
        const dep = provider.deps[args.index] || { id: null };
        provider.deps[args.index] = {
            id: args.injectable ? this.providerId(args.injectable) : dep.id,
            optional: args.optional || dep.optional,
        };
        return type;
    }
    /**
     * Find stored provider
     */
    static findProvider(injectable) {
        const id = this.providerId(injectable);
        return this.providers.find((provider) => provider.id === id);
    }
    /**
     * Replace stored provider wiht new provider
     */
    static replaceProvider(injectable, provider) {
        const storeProvider = this.findProvider(injectable);
        const index = this.providers.indexOf(storeProvider);
        if (index !== -1) {
            this.providers[index] = provider;
        }
        else {
            this.providers.push(provider);
        }
    }
    /**
     * Create and store provider
     */
    static createProvider(type) {
        const id = reflector_1.Reflector.setId(type);
        const storeProvider = { id, type, deps: [] };
        this.providers.push(storeProvider);
        return storeProvider;
    }
}
exports.Store = Store;
Store.providers = [];
