// @decorators/express/lib/src/middleware.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorMiddlewareHandler = exports.ERROR_MIDDLEWARE = exports.middlewareHandler = void 0;
const di_1 = require("@decorators/di");
/**
 * Create request middleware handler that uses class or function provided as middleware
 */
function middlewareHandler(middleware) {
    return (req, res, next) => {
        invokeMiddleware(middleware, [req, res, next]).catch(next);
    };
}
exports.middlewareHandler = middlewareHandler;
/**
 * Error Middleware class registration DI token
 */
exports.ERROR_MIDDLEWARE = new di_1.InjectionToken('ERROR_MIDDLEWARE');
/**
 * Add error middleware to the app
 */
function errorMiddlewareHandler() {
    return (error, req, res, next) => {
        invokeMiddleware(exports.ERROR_MIDDLEWARE, [error, req, res, next]).catch(next);
    };
}
exports.errorMiddlewareHandler = errorMiddlewareHandler;
/**
 * Instantiate middleware and invoke it with arguments
 */
function invokeMiddleware(middleware, args) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        const next = args[args.length - 1];
        try {
            let instance;
            if (typeof middleware === 'function') {
                if ((_a = middleware.prototype) === null || _a === void 0 ? void 0 : _a.use) {
                    instance = new middleware(...args);
                }
                else {
                    instance = middleware;
                }
            }
            else {
                instance = yield di_1.Container.get(middleware);
            }
            const handler = (_b = instance === null || instance === void 0 ? void 0 : instance.use) !== null && _b !== void 0 ? _b : instance;
            const result = typeof handler === 'function' ? handler.apply(instance, args) : instance;
            if (result instanceof Promise) {
                result.catch(next);
            }
        }
        catch (err) {
            next(err);
        }
    });
}
